### PR TITLE
FAI-287: Refactor embedding of DMN in trusty-ai

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/FAI-287

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

You can run this PR by:-

- Run this code 
```
cd ./kogito-apps/ui-packages/packages/trusty
yarn run mock-server

[new window]
yarn run start
```
- Click on the first thing in the list
- Click on the Model Lookup" tab

### CODE REVIEW

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx [hunk 1]
```diff
@@ -14,49 +12,26 @@ const ModelDiagram = (props: ModelDiagramProps) => {
...
+  useEffect(() => {
+    if (type === DMN1_2) {
+      DmnEditor.open({
+        container: document.getElementById('dmn-editor-container'),
+        initialContent: Promise.resolve(model.model),
+        readOnly: false,
+        origin: ''
+      });
+    }
+  }, [model]);
```
Comment (tiagobento) [inline ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: You need a cleanup bit here as well. Capture the Editor instance created by `open` and call a `editor.close()` on the returning lambda of this effect.

**Comment by manstis** [comment]
@tiagobento Updated. Thank-you. Also, removing `origin` as you proposed fixes embedding too. Perfect.

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/tests/ModelDiagram.test.tsx [hunk 2]
```diff
@@ -3,12 +3,17 @@ import { mount } from 'enzyme';
 import ModelDiagram from '../ModelDiagram';
 import { ModelData } from '../../../../types';
 
+const mockOpenFunction = jest.fn();
+jest.mock('@kogito-tooling/kie-editors-standalone/dist/dmn', () => ({
+  open: () => mockOpenFunction()
+}));
+
```
Comment (ljmotta) [inline ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/tests/ModelDiagram.test.tsx:10]: Is it not necessary to reset the mock after the test suit with a `jest.resetAllMocks()` on the `afterAll` callback? Depending on the test order, this mock can influence another test file.

File: ui-packages/packages/trusty/src/components/Templates/ModelLookup/tests/ModelLookup.test.tsx [hunk 3]
```diff
@@ -16,6 +16,7 @@ jest.mock('react-router-dom', () => ({
     url: '/audit/:executionType/:executionId/model-lookup'
   })
 }));
+jest.mock('@kogito-tooling/kie-editors-standalone/dist/dmn');
```
Comment (ljmotta) [inline ui-packages/packages/trusty/src/components/Templates/ModelLookup/tests/ModelLookup.test.tsx:19]: The same question about the `jest.resetAllMocks()`. If any other component on another test file uses any of these hooks, will they not use this mock?

**Review by ljmotta — COMMENTED** [review]
Hi @manstis . I've made some comments inline.

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/tests/ModelDiagram.test.tsx (hunk 2 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/tests/ModelDiagram.test.tsx:10]: Sure thing. Added.

File: ui-packages/packages/trusty/src/components/Templates/ModelLookup/tests/ModelLookup.test.tsx (hunk 3 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Templates/ModelLookup/tests/ModelLookup.test.tsx:19]: Sure thing. Added.

**Comment by manstis** [comment]
@tiagobento @ljmotta Updates completed. Please can you re-review/approve?

**Comment by danielezonca** [comment]
@manstis 
Please update README too
https://github.com/kiegroup/kogito-apps/blob/master/ui-packages/packages/trusty/README.md#install-gwt-editors

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx [hunk 4]
```diff
@@ -14,49 +13,31 @@ const ModelDiagram = (props: ModelDiagramProps) => {
...
+        container: document.getElementById('dmn-editor-container'),
+        initialContent: Promise.resolve(model.model),
+        readOnly: true
+      });
+    }
+    return () => {
+      if (editor) {
+        editor.close();
+      }
+    };
```
Comment (tiagobento) [inline ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: You can actually use return the cleanup function of this `useEffect` inside the if. The reasoning is that when the Editor is created, we tell React to close it when cleaning up. When the Editor is not created (`type !== DMN1_2`), we don't need to cleanup anything. So:
```suggestion
    if (type === DMN1_2) {
      const editor = DmnEditor.open({
        container: document.getElementById('dmn-editor-container'),
        initialContent: Promise.resolve(model.model),
        readOnly: true
      });
      
      return () => {
        editor.close();
      }
    }
```

Comment (tiagobento) [reply ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: Approving as this is optional.

**Review by tiagobento — APPROVED** [review]

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx (hunk 4 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: @tiagobento Update made. Thank-you (and thank-you for approving).

**Comment by manstis** [comment]
Jenkins retest this

**Review by ljmotta — APPROVED** [review]

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
```
[INFO] Running org.kie.kogito.jitexecutor.dmn.JITDMNServiceImplTest

[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.2 s - in org.kie.kogito.jitexecutor.dmn.JITDMNServiceImplTest

[INFO] Running org.kie.kogito.jitexecutor.dmn.api.ValidatorResourceTest

Feb 23, 2021 10:41:46 AM org.jboss.threads.Version <clinit>

INFO: JBoss Threads version 3.2.0.Final

[ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 6.274 s <<< FAILURE! - in org.kie.kogito.jitexecutor.dmn.api.ValidatorResourceTest

[ERROR] testOverlap  Time elapsed: 0.015 s  <<< ERROR!

java.lang.RuntimeException: java.lang.NoSuchMethodException: io.quarkus.test.common.TestResourceManager.<init>(java.lang.Class, java.util.List)

	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:574)

	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:644)

	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)

	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)

	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:72)

	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)

	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)

	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)

	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)

	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)

	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)

	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:77)

	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:342)

	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:289)

	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:79)

	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:267)

	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$2(ClassBasedTestDescriptor.java:259)
```
@r00ta @tarilabs IDK if this is a false negative or something someone needs to be concerned about?

**Comment by manstis** [comment]
Jenkins retest this

**Comment by tarilabs** [comment]
weird @manstis , seems as internal failure of the Quarkus test infra.. something maybe related to the Quarkus update?

**Comment by manstis** [comment]
@tarilabs Do you know who might be interested in knowing? 

If the failure persists I'd like to alert anybody who may have an interest in it. As a last resort I'll email BSIG should it continue to fail.

**Comment by tarilabs** [comment]
It's @r00ta and myself, I'm only saying, that seems an internal issue of the Quarkus test infra, and since the test was passing earlier, the only thing I could think of is that this might have been in connection to the Quarkus update 🤷‍♂️

**Comment by tarilabs** [comment]
One more information about  https://github.com/kiegroup/kogito-apps/pull/654#issuecomment-784302336

I've checked [here](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/kogito/job/pullrequest/job/kogito-apps/171/#showFailuresLink)

It seems to me ALL of these tests:
```
org.kie.kogito.index.service.InfinispanIndexingServiceIT.org.kie.kogito.index.service.InfinispanIndexingServiceIT
org.kie.kogito.index.service.InfinispanKeycloakIntegrationIndexingServiceIT.testUnauthorizedUserAccess
org.kie.kogito.explainability.messaging.ExplainabilityMessagingHandlerIT.explainabilityRequestIsProcessedAndAResultMessageIsSent
org.kie.kogito.explainability.rest.ExplainabilityApiV1Test.testEndpointWithRequest
org.kie.kogito.jitexecutor.dmn.api.ValidatorResourceTest.testOverlap
org.kie.kogito.jobs.service.repository.infinispan.InfinispanJobRepositoryIT.org.kie.kogito.jobs.service.repository.infinispan.InfinispanJobRepositoryIT
org.kie.kogito.jobs.service.resource.KeycloakJobServiceIT.org.kie.kogito.jobs.service.resource.KeycloakJobServiceIT
org.kie.kogito.persistence.mongodb.HealthCheckIT.testHealthCheck
org.kie.kogito.security.KeycloakSecurityCommonsServiceIT.meTest
org.kie.kogito.taskassigning.index.service.client.graphql.GraphQLServiceClientFactoryTest.newClient
org.kie.kogito.trusty.ui.StaticContentTest.testIndexHtml
org.kie.kogito.trusty.service.common.KeycloakTrustyServiceIT.org.kie.kogito.trusty.service.common.KeycloakTrustyServiceIT
org.kie.kogito.trusty.service.common.messaging.incoming.ModelEventConsumerIT.eventLoopIsNotStoppedWithException
```

are failing for the same stacktrace, proof:
image

So this would, to me, lead to the conclusion that is nothing specifically about the jitrunner,
but is a more generalized issue either on this PR or on kogito-apps.

Since recently there has been some quarkus update, I would try to rebase the PR on top of the most recent changes, but again based on this evidence I would not conclude is something specific to the jitrunner.

**Review by sjamboro — APPROVED** [review]

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx [hunk 5]
```diff
@@ -14,49 +13,29 @@ const ModelDiagram = (props: ModelDiagramProps) => {
...
+        editor.close();
+      };
+    }
+  }, [model]);
+
+  return type === DMN1_2 ? makeDMNEditor() : DEFAULT;
 };
 
 function makeUnknownModel(): JSX.Element {
   return <div>Unknown model type</div>;
```
Comment (kelvah) [inline ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: Not related to the PR but I would improve a bit the unknown model type message. Like adding an [EmptyState](https://www.patternfly.org/2020.16/components/empty-state).

**Review by kelvah — APPROVED** [review]
Looking good @manstis. Added a comment for an improvement if you feel like to.

File: ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx (hunk 5 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Organisms/ModelDiagram/ModelDiagram.tsx]: New commit added for `EmptyState`.

**Pull request merged**